lib/puppet/network/http/api/indirected_routes.rb

```diff
@@ -190,9 +190,21 @@ def accepted_response_formatter_or_json_for(model_class, request)
 
   def read_body_into_model(model_class, request)
     data = request.body.to_s
+    formatter = request.formatter
+
+    if formatter && formatter.supported?(model_class)
+      begin
+        return model_class.convert_from(formatter.name.to_s, data)
+      rescue => e
+        raise Puppet::Network::HTTP::Error::HTTPBadRequestError.new(
+          _("The request body is invalid: %{message}") % { message: e.message })
+      end
+    end
 
-    format = request.format
-    model_class.convert_from(format, data)
+    #TRANSLATORS "mime-type" is a keyword and should not be translated
+    raise Puppet::Network::HTTP::Error::HTTPUnsupportedMediaTypeError.new(
+      _("Client sent a mime-type (%{header}) that doesn't correspond to a format we support") % { header: request.headers['content-type'] },
+      Puppet::Network::HTTP::Issues::UNSUPPORTED_MEDIA_TYPE)
```

Comment:
Good question. The `IndirectedRoutes` class only handles a subset of REST requests, and there are other routes for things that aren't indirected types, e.g. `/puppet/v3/environments`. While it appears we only examine the request body when `PUT`ing an indirected model, e.g. CSR, Report, it's not safe to assume that will always be the case.

If the agent sends an unknown `Content-Type`, then we should handle that more generally in `Request#formatter` as that logic is shared for all routes and all servers (webrick/puppetserver/rack). Then I can leave this check for indirected models that don't support the requested format, e.g. uploading a report in `application/octet-stream`.